linkerd/app/outbound/src/tls/detect.rs

```diff
@@ -0,0 +1,92 @@
+use linkerd_app_core::{
+    io,
+    svc::{self, layer, ServiceExt},
+    tls::{
+        server::{detect_sni, DetectIo},
+        ServerName,
+    },
+    Error, Result,
+};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use thiserror::Error;
+use tokio::time;
+
+#[derive(Clone, Debug, Error)]
+#[error("SNI detection timed out")]
+pub struct SniDetectionTimeoutError;
+
+#[derive(Clone, Debug, Error)]
+#[error("Could not find SNI")]
+pub struct NoSniFoundError;
+
+#[derive(Clone, Debug, Default)]
+pub struct NewDetectSniServer<N> {
+    inner: N,
+    timeout: time::Duration,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct DetectSniServer<T, N> {
```

Comment:
How does this differ from the middleware in linkerd/tls/src/server.rs? Can they be unified? Is there any reason that this middleware can't be defined in the linkerd-tls crate? That should be our default--it's easier to to test/change modules when they are decoupled from the app. That is, if the app isn't compiling, it can still be possible to test and make changes to this stuff.